### PR TITLE
Check ssl->arrays in SendClientHello to avoid null dereference.  Allow building with fallthrough defined.

### PR DESCRIPTION
**Pull request by kareem-wolfssl**
GH #4568
ZD 13238

### CODE REVIEW

File: wolfssl/wolfcrypt/types.h [hunk 1]
```diff
@@ -315,7 +315,8 @@ decouple library dependencies with standard string, memory and so on.
                 #if defined(WOLFSSL_LINUXKM) && defined(fallthrough)
                     #define FALL_THROUGH fallthrough
                 #else
-                    #define FALL_THROUGH ; __attribute__ ((fallthrough))
+                    /* Use __ notation to avoid conflicts */
```
Comment (dgarske) [inline wolfssl/wolfcrypt/types.h:318]: Another option would be to use `fallthrough` if it is already defined, just like the Linux KM does above.

```
#ifdef fallthrough
    #define FALL_THROUGH fallthrough
#else
    #define FALL_THROUGH ; __attribute__ ((__fallthrough__))
#endif
```

@douzzer , what do you think?

**Review by dgarske — CHANGES_REQUESTED** [review]

File: wolfssl/wolfcrypt/types.h (hunk 1 above)
Comment (kareem-wolfssl) [reply wolfssl/wolfcrypt/types.h:318]: Agreed that is smarter, in case the platform defines fallthrough to something else.  Will update.

Comment (douzzer) [reply wolfssl/wolfcrypt/types.h:318]: yup, can't see anything wrong with
```
#if defined(__GNUC__) && ((__GNUC__ > 7) || ((__GNUC__ == 7) && (__GNUC_MINOR__ >= 1)))
#ifdef fallthrough
#define FALL_THROUGH fallthrough
#else
#define FALL_THROUGH ; __attribute__ ((__fallthrough__))
#endif
#endif
```
am currently experimenting with how to make `clang` happy on this stuff too.  there's a `-Wimplicit-fallthrough` that's disabled by default.

Comment (douzzer) [reply wolfssl/wolfcrypt/types.h:318]: here we go, this does the trick:
```
diff --git a/wolfssl/wolfcrypt/types.h b/wolfssl/wolfcrypt/types.h
index 0d18df297..5beb7c022 100644
--- a/wolfssl/wolfcrypt/types.h
+++ b/wolfssl/wolfcrypt/types.h
@@ -311,12 +311,13 @@ decouple library dependencies with standard string, memory and so on.
     #ifndef FALL_THROUGH
         /* GCC 7 has new switch() fall-through detection */
         #if defined(__GNUC__)
-            #if ((__GNUC__ > 7) || ((__GNUC__ == 7) && (__GNUC_MINOR__ >= 1)))
-                #if defined(WOLFSSL_LINUXKM) && defined(fallthrough)
-                    #define FALL_THROUGH fallthrough
-                #else
-                    #define FALL_THROUGH ; __attribute__ ((fallthrough))
-                #endif
+            #if defined(fallthrough)
+                #define FALL_THROUGH fallthrough
+            #elif ((__GNUC__ > 7) || ((__GNUC__ == 7) && (__GNUC_MINOR__ >= 1)))
+                #define FALL_THROUGH ; __attribute__ ((fallthrough))
+            #elif defined(__clang__) && defined(__clang_major__) &&
+                    (__clang_major__ >= 4)
+                #define FALL_THROUGH ; __attribute__ ((fallthrough))
             #endif
         #endif
     #endif /* FALL_THROUGH */
```

Comment (kareem-wolfssl) [reply wolfssl/wolfcrypt/types.h]: Updated, thanks Douzzer!

Comment (douzzer) [reply wolfssl/wolfcrypt/types.h]: whoops @kareem-wolfssl , I didn't test what I posted above!  I forgot the backslash needed here:
```
            #elif defined(__clang__) && defined(__clang_major__) && \
                    (__clang_major__ >= 4)
```
(I "fixed" an overlong line _after_ testing...)

File: src/internal.c [hunk 2]
```diff
@@ -22665,6 +22665,10 @@ int DecodePrivateKey(WOLFSSL *ssl, word16* length)
         WOLFSSL_START(WC_FUNC_CLIENT_HELLO_SEND);
         WOLFSSL_ENTER("SendClientHello");
 
+        if (ssl == NULL || ssl->arrays == NULL) {
+            return BAD_FUNC_ARG;
+        }
```
Comment (mstaz) [inline src/internal.c:22670]: Is there a case where ClientHello should be sent although `ssl->arrays` is NULL? E.g. when `ssl->options.connectState` is not `CONNECT_BEGIN` and DTLS is disabled. However don't know if that makes sense?

Comment (kareem-wolfssl) [reply src/internal.c]: Yes, potentially.  Updated this check.

File: wolfssl/wolfcrypt/types.h (hunk 1 above)
Comment (kareem-wolfssl) [reply wolfssl/wolfcrypt/types.h]: Updated.

**Review by dgarske — APPROVED** [review]

**Pull request merged**